### PR TITLE
Normalize Etag format

### PR DESCRIPTION
**Pull request by wonwuakpa-msft**
## Description 

This fixes a bug with the Etag blob property where `info()` and `ls()` were returning different formats. This [GitHub issue](https://github.com/fsspec/adlfs/issues/461) reproduces said bug. This is due to API differences when using `BlobClient.get_blob_properties()` vs  `BlobContainer.walk_blobs()`. 
This PR introduces a standardization to `_details()` by ensuring the Etag property return type is wrapped with double quotes.  S3fs formats Etags with double quotes [here](https://github.com/fsspec/s3fs/blob/731e1250bcd4f682e1ccce03b01641910e7646fc/s3fs/core.py#L1741)

## Test
Added `test_etag_normalized_form` to test both calls property is equivalent and there is no double quoting

### CODE REVIEW

File: adlfs/spec.py [hunk 1]
```diff
@@ -897,6 +897,12 @@ async def _details(
                 for key in FORWARDED_BLOB_PROPERTIES
                 if content.has_key(key)  # NOQA
             }
+
+            # Normalize etag to always return a string with quotes for consistency
+            if data.get("etag") and data["etag"] is not None:
```
Comment (nateprewitt) [inline adlfs/spec.py]: I think we're checking the same thing twice here, so we may be able to simplify. `get()` will give us back `None` if it doesn't exist so we can just check the value there. These are a couple alternatives we can use depending on how we want to mutate the variable. More on that below.

```suggestion
            if data.get("etag") is not None:
```
or
```suggestion
            if (etag_val := data.get("etag")) is not None:
```

If we're concerned about the server sending us back an empty string (that shouldn't happen?), we can change our default. This validates the value is "truthy" so we're not quoting an empty string. I'm not actually sure what the desired behavior is here though, I don't think that'd ever be a valid etag.

```suggestion
            if (etag_val := data.get("etag", "")):
```

File: adlfs/spec.py [hunk 2]
```diff
@@ -897,6 +897,12 @@ async def _details(
...
+                etagVal = data["etag"]
```
Comment (nateprewitt) [inline adlfs/spec.py]: Python has a lot of semantics built into casing for variables. [PEP 8](https://peps.python.org/pep-0008/#prescriptive-naming-conventions) codifies most of those and is a really useful reference when in doubt. Function scope variable names should be snake_case (i.e. `etag_val` in this case).

File: adlfs/spec.py [hunk 3]
```diff
@@ -897,6 +897,12 @@ async def _details(
...
+                if not etagVal.startswith('"') and not etagVal.endswith('"'):
+                    data["etag"] = f'"{etagVal}"'
```
Comment (nateprewitt) [inline adlfs/spec.py]: I'm not sure this covers all problematic cases. It's unlikely we'll ever get one of these back from Azure, but probably best to be uniform with how we apply this fix. Values like `1234"` will pass this silently.

A more robust option may be to use `strip("\"")` to remove quotes on the edges and then force requoting each time. Something like the code below.

```suggestion
                data["etag"] = f'"{data["etag"].strip("\"")}"'
```
or if we kept the variable from the walrus operator above:
```suggestion
                etag_val = etag_val.strip("\"")
                data["etag"] = f'"{etag_val}"'
```

File: adlfs/tests/test_spec.py [hunk 4]
```diff
@@ -2571,3 +2571,28 @@ class TestCloseCredential:
...
+    info_etag = fs.info(path, refresh=True)["etag"]
+
+    assert info_etag == ls_etag
+
+    # Validate both etags are quoted
+    assert ls_etag.startswith('"') and info_etag.startswith('"')
+    assert ls_etag.endswith('"') and info_etag.endswith('"')
+
+    # Validate etag from info() is not double quoted
+    assert not info_etag.startswith('""') and not info_etag.endswith('""')
```
Comment (nateprewitt) [inline adlfs/tests/test_spec.py]: nit; but since we're comparing the etags we can validate one and then the comparison ensures we have everything inline.
```suggestion
    # Validate etag is quoted but only once.
    assert ls_etag.startswith('"') and ls_etag.endswith('"')
    assert not ls_etag.startswith('""') and not ls_etag.endswith('""')
    
    # Validate etags are identical
    assert info_etag == ls_etag
```

File: adlfs/tests/test_spec.py [hunk 5]
```diff
@@ -2571,3 +2571,28 @@ class TestCloseCredential:
...
+
+def test_etag_normalized_form(storage):
+    """
+    Tests a consistent quoted etag format with ls() and info() calls.
+    """
+    fs = AzureBlobFileSystem(
+        account_name=storage.account_name,
+        connection_string=CONN_STR,
+    )
+    path = "data/root/a/file.txt"
```
Comment (nateprewitt) [inline adlfs/tests/test_spec.py:2585]: It looks like we're getting our etag back from Azurite checksumming the file? That seems reasonable for an e2e test, but I'm wondering if we want to unit test the normalization behavior any more for edge cases like an empty value (`""`) or lop-sided quotes to make sure we don't fail unexpectedly. Uniformity is good but we don't want to introduce runtime failures for it.

We could break the normalization into a testable function if needed. I'm not sure if @kyleknap or @anjaliratnam-msft  have an opinion.

**Review by nateprewitt — CHANGES_REQUESTED** [review]
Thanks for the PR, @wonwuakpa-msft, this is a great start! I left some initial feedback on a few things. Please let me know if you think I'm missing something or if you have any questions.

File: adlfs/spec.py (hunk 1 above)
Comment (wonwuakpa-msft) [reply adlfs/spec.py]: Yea, I can simplify this validation to use just the getter. I agree the service can only send us back either None (when the header isn't set) or a valid Etag string. So, there is no need to handle an empty string.

File: adlfs/tests/test_spec.py (hunk 5 above)
Comment (anjaliratnam-msft) [reply adlfs/tests/test_spec.py:2585]: Yeah, I agree. I think it could be worth it to add unit tests just to make sure it's always gets wrapped with double quotes.

File: adlfs/tests/test_spec.py [hunk 6]
```diff
@@ -2571,3 +2571,33 @@ class TestCloseCredential:
...
+
+    assert info_etag == ls_etag
+
+    # Validate both etags are quoted
+    assert ls_etag.startswith('"') and info_etag.startswith('"')
+    assert ls_etag.endswith('"') and info_etag.endswith('"')
+
+    # Validate etag from info() is not double quoted
+    # Validate etag is quoted but only once.
+    assert ls_etag.startswith('"') and ls_etag.endswith('"')
```
Comment (anjaliratnam-msft) [inline adlfs/tests/test_spec.py]: I think you can get rid of this line since it's already being checked by 2594 and 2595.

File: adlfs/spec.py [hunk 7]
```diff
@@ -1,2337 +1,2363 @@
...
+
+def _create_aio_blob_service_client_from_connection_string(
+    connection_string: str,
+) -> AIOBlobServiceClient:
+    return AIOBlobServiceClient.from_connection_string(
+        conn_str=connection_string,
+        user_agent=_USER_AGENT,
+    )
+
+
```
Comment (wonwuakpa-msft) [inline adlfs/spec.py]: `_strip_quotes_from_etag` and its calling location is the only addition in this file. Not sure if this is the best place to add this helper. It makes the diff really large

**Comment by nateprewitt** [comment]
@wonwuakpa-msft It looks like the latest commit may have modified the newline at the end of spec.py so it's diffing the entire file. Can you take a look at that? Feel free to ping me if you need help.

**Comment by wonwuakpa-msft** [comment]
> @wonwuakpa-msft It looks like the latest commit may have modified the newline at the end of spec.py so it's diffing the entire file. Can you take a look at that? Feel free to ping me if you need help.

Yes, it was changing the line endings of every line in spec.py from LF to CRLF, so git detected that as a new change. Setting the `git config core.autocrlf = input` and changing my settings.json file in VSCode to `/n` resolved it.

File: adlfs/tests/test_spec.py [hunk 8]
```diff
@@ -2571,3 +2572,41 @@ class TestCloseCredential:
...
+    assert ls_etag.startswith('"') and info_etag.startswith('"')
+    assert ls_etag.endswith('"') and info_etag.endswith('"')
+    # Validate etag is not double quoted
+    assert not ls_etag.startswith('""') and not ls_etag.endswith('""')
+
+
+@pytest.mark.parametrize(
+    "input_etag,expected_etag",
+    [
+        pytest.param("0xA123456", '"0xA123456"', id="bug"),
```
Comment (anjaliratnam-msft) [inline adlfs/tests/test_spec.py]: Instead of doing pytest.param, I think you can just put the values for input_etag and expected_etag, similar to the other tests. I don't really think an id is necessary for these.

**Review by anjaliratnam-msft — COMMENTED** [review]
This looks good to me, I just had one small comment. Other than that, I think you should be able to rebase off of main as well, since there have been some updates there.

File: adlfs/spec.py [hunk 9]
```diff
@@ -156,6 +156,27 @@ def _create_aio_blob_service_client_from_connection_string(
...
+    Parameters
+    ----------
+    etag: str
+        Raw etag value from Azure
+
+    Returns
+    -------
+    str
+        Returns the normalized etag
+    """
```
Comment (nateprewitt) [inline adlfs/spec.py]: nit; Generally, docstrings aren't needed for private functions unless there's a lot of nuance since end-users won't interact with them.

File: adlfs/spec.py [hunk 10]
```diff
@@ -156,6 +156,27 @@ def _create_aio_blob_service_client_from_connection_string(
     )
 
 
+def _strip_quotes_from_etag(etag: str) -> Optional[str]:
```
Comment (nateprewitt) [inline adlfs/spec.py]: We're quoting the etag so something like this may be clearer about intent.
```suggestion
def _normalize_etag_quotes(etag: str) -> Optional[str]:
```

**Review by nateprewitt — COMMENTED** [review]
Looks good! A couple minor comments and once @anjaliratnam-msft's comments are addressed we should be set.

File: adlfs/tests/test_spec.py [hunk 11]
```diff
@@ -2571,3 +2572,33 @@ class TestCloseCredential:
...
+    info_etag = fs.info(path, refresh=True)["etag"]
+
+    assert info_etag == ls_etag
+    # Validate both etags are quoted
+    assert ls_etag.startswith('"') and info_etag.startswith('"')
+    assert ls_etag.endswith('"') and info_etag.endswith('"')
+    # Validate etag is not double quoted
+    assert not ls_etag.startswith('""') and not ls_etag.endswith('""')
+
+    assert _normalize_etag_quotes("0xA123456") == '"0xA123456"'
```
Comment (anjaliratnam-msft) [inline adlfs/tests/test_spec.py]: Oh sorry, I just meant removing the pytest.param part but still keeping it under the parametrization under a separate test. something like this
```suggestion
    @pytest.mark.parametrize(
    "input_etag,expected_etag",
    [
        ("0xA123456", '"0xA123456"')
        ('"0xA123456"', '"0xA123456"'),
        ('""0xA123456""', '"0xA123456"'),
        ('"0xA123456', '"0xA123456"'),
        ('0xA123456"', '"0xA123456"'),
        (None, None),
    ],
```

**Review by nateprewitt — APPROVED** [review]
:shipit: Thank you @wonwuakpa-msft!

**Review by kyleknap — APPROVED** [review]
Looks good to me! :ship:
